### Importing Pandas library and set as "pd"

In [1]:
import pandas as pd
import datetime
%matplotlib inline

## Section 4.2: Read in your data

Data file salaries.csv was used for this exercise. Data source is Kaggle: https://www.kaggle.com/datasets/arnabchaki/data-science-salaries-2023 

Column names are created, and set either, integer, category or float data types, depending on the underlying data.
Spaces are replaced with Underscores. The created Dtypes are displayed below in the output

In [2]:
salary = pd.read_csv('../data/ds_salaries.csv',  
                     dtype={'work_year': 'object',
                           'experience_level': 'category',                                                    
                           'employment_type': 'category',
                           'job_title': 'category',
                           'salary': 'float64',
                           'salary_currency': 'category',
                           'job_title': 'category',
                           'salary_in_usd': 'float64',
                           'employee_residence': 'category',
                           'remote_ratio': 'object',
                           'company_location': 'category',                                                      
                           'company_size': 'category',},
                           nrows=5000,
                    )
salary.rename(columns=lambda x: x.strip(" ").replace(' ', '_'), inplace=True)
salary.dtypes

work_year               object
experience_level      category
employment_type       category
job_title             category
salary                 float64
salary_currency       category
salary_in_usd          float64
employee_residence    category
remote_ratio            object
company_location      category
company_size          category
dtype: object

## Section 4.3: Validating the packaging

In [3]:
print(salary.shape)
print("This data has", len(salary.index),"Rows and", len(salary.columns), "columns" )

(3755, 11)
This data has 3755 Rows and 11 columns


In [4]:
##Decribe() function does a statistical analysis on the data.
salary.describe(include='all').T

count unique            top  freq           mean   
work_year             3755      4           2023  1785            NaN  \
experience_level      3755      4             SE  2516            NaN   
employment_type       3755      4             FT  3718            NaN   
job_title             3755     93  Data Engineer  1040            NaN   
salary              3755.0    NaN            NaN   NaN  190695.571771   
salary_currency       3755     20            USD  3224            NaN   
salary_in_usd       3755.0    NaN            NaN   NaN   137570.38988   
employee_residence    3755     78             US  3004            NaN   
remote_ratio          3755      3              0  1923            NaN   
company_location      3755     72             US  3040            NaN   
company_size          3755      3              M  3153            NaN   

                              std     min       25%       50%       75%   
work_year                     NaN     NaN       NaN       NaN       NaN  \
experience_level              NaN     NaN       NaN       NaN       NaN   
employment_type               NaN     NaN       NaN       NaN       NaN   
job_title                     NaN     NaN       NaN       NaN       NaN   
salary              671676.500508  6000.0  100000.0  138000.0  180000.0   
salary_currency               NaN     NaN       NaN       NaN       NaN   
salary_in_usd        63055.625278  5132.0   95000.0  135000.0  175000.0   
employee_residence            NaN     NaN       NaN       NaN       NaN   
remote_ratio                  NaN     NaN       NaN       NaN       NaN   
company_location              NaN     NaN       NaN       NaN       NaN   
company_size                  NaN     NaN       NaN       NaN       NaN   

                           max  
work_year                  NaN  
experience_level           NaN  
employment_type            NaN  
job_title                  NaN  
salary              30400000.0  
salary_currency            NaN  
salary_in_usd         450000.0  
employee_residence         NaN  
remote_ratio               NaN  
company_location           NaN  
company_size               NaN

In [5]:
salary.sample(5).T

793            35     
work_year                           2023           2023  \
experience_level                      SE             MI   
employment_type                       FT             FT   
job_title           Data Science Manager  Data Engineer   
salary                          299500.0       162500.0   
salary_currency                      USD            USD   
salary_in_usd                   299500.0       162500.0   
employee_residence                    US             US   
remote_ratio                           0              0   
company_location                      US             US   
company_size                           M              M   

                                                  2818           2433   
work_year                                         2022           2022  \
experience_level                                    SE             SE   
employment_type                                     FT             FT   
job_title           Applied Machine Learning Scientist  Data Engineer   
salary                                        108000.0       184100.0   
salary_currency                                    USD            USD   
salary_in_usd                                 108000.0       184100.0   
employee_residence                                  US             US   
remote_ratio                                         0              0   
company_location                                    US             US   
company_size                                         L              M   

                           1526  
work_year                  2023  
experience_level             SE  
employment_type              FT  
job_title           ML Engineer  
salary                 135000.0  
salary_currency             USD  
salary_in_usd          135000.0  
employee_residence           US  
remote_ratio                  0  
company_location             US  
company_size                  M

## Section 4.4: Look at the top and the bottom

The top 5 rows were validated against the csv file, and is correct

In [6]:
##returning the top 5 rows
salary.iloc[:,[1,3,0]].head()

experience_level                 job_title work_year
0               SE  Principal Data Scientist      2023
1               MI               ML Engineer      2023
2               MI               ML Engineer      2023
3               SE            Data Scientist      2023
4               SE            Data Scientist      2023

The bottom 5 rows were validated against the csv file, and is correct

In [7]:
##returning the bottom 5 rows
salary.iloc[:,[1,3,0]].tail()

experience_level                 job_title work_year
3750               SE            Data Scientist      2020
3751               MI  Principal Data Scientist      2021
3752               EN            Data Scientist      2020
3753               EN     Business Data Analyst      2020
3754               SE      Data Science Manager      2021

## Section 4.5: ABC: Always be Checking Your “n”s

Majority of the candidates (67%) were seniors (SE) and (21%) Mid/intermediate level (MI). So expecting to see the composition containing at least these two values.

In [8]:
salary.experience_level.head(20)

0     SE
1     MI
2     MI
3     SE
4     SE
5     SE
6     SE
7     SE
8     SE
9     SE
10    SE
11    SE
12    SE
13    EN
14    EN
15    SE
16    SE
17    SE
18    SE
19    MI
Name: experience_level, dtype: category
Categories (4, object): ['EN', 'EX', 'MI', 'SE']

There are only 4 types allowed employment types, being Full tume (FT), Part time (PT), Freelancer (FL) and Contractor (CT)

In [9]:
salary.employment_type.unique()

['FT', 'CT', 'FL', 'PT']
Categories (4, object): ['CT', 'FL', 'FT', 'PT']

There are only 4 allowed experience levels, being (SE) Senior, (EN)Entry level, (EX) Executive level, and (MI) Mid/Intermediate level

In [10]:
salary.experience_level.unique()

['SE', 'MI', 'EN', 'EX']
Categories (4, object): ['EN', 'EX', 'MI', 'SE']

An extract of data scientists that are Full time earning a salary above $300 000 in most recent year of 2023. Expecting 5 results.

In [11]:
salary.loc[(salary.employment_type=='FT') & 
           (salary.job_title=='Data Scientist') &
          (salary.salary >= 300000) &
          (salary.salary_currency == 'USD') &
          (salary.work_year >= '2023'),
          ['employment_type','job_title','salary','salary_currency','work_year']]

employment_type       job_title    salary salary_currency work_year
488               FT  Data Scientist  317070.0             USD      2023
1097              FT  Data Scientist  300240.0             USD      2023
1099              FT  Data Scientist  300240.0             USD      2023
1105              FT  Data Scientist  370000.0             USD      2023
1605              FT  Data Scientist  300000.0             USD      2023

In [12]:
len(salary.job_title.unique())

93

In [13]:
pd.Series(salary.salary_currency.cat.categories)

0     AUD
1     BRL
2     CAD
3     CHF
4     CLP
5     CZK
6     DKK
7     EUR
8     GBP
9     HKD
10    HUF
11    ILS
12    INR
13    JPY
14    MXN
15    PLN
16    SGD
17    THB
18    TRY
19    USD
dtype: object

## Section 4.6: Validate With at Least One External Data Source

The sample size contains majority Senior individuals. 4 different expereince levels are avaliable. SE roles appeared 2516 of the total sample which is approximately 67%.

In [14]:
salary.experience_level.describe()

count     3755
unique       4
top         SE
freq      2516
Name: experience_level, dtype: object

The salary values were sampled and grouped into different quantiles. Majority of roles within the 4th till 8th quantile.

In [15]:
salary.salary.quantile([0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0])

0.0        6000.0
0.1       60000.0
0.2       85000.0
0.3      106260.0
0.4      123628.8
0.5      138000.0
0.6      150000.0
0.7      170000.0
0.8      191475.0
0.9      228600.0
1.0    30400000.0
Name: salary, dtype: float64

## Section 4.7

In [16]:
import seaborn as sns

In [ ]:
g = sns.catplot(data=salary.sample(3755), x='job_title', y='salary_in_usd', kind='box', aspect=3.0)
g.set_axis_labels("", "Salary (USD)")
g.set_xticklabels(rotation=90);

TypeError: Neither the `x` nor `y` variable appears to be numeric.

### Try the Easy Solution First

In [33]:
salary['salary_in_usd'] = salary.salary_in_usd.map(lambda x: 'high' if x < -100 else 'low')

In [35]:
salary.groupby(by='salary_in_usd').salary_in_usd.describe()

count unique  top  freq
salary_in_usd                        
low            3755      1  low  3755

In [ ]:
g = sns.catplot(data=salary.sample(1000000), x='Region', y='Sample_Measurement', kind='box', aspect=2.0)
g.set_axis_labels("", "Ozone level (ppm)")
g.set_xticklabels(rotation=90);

#### Challenge Your Solution

In [ ]:
salary.loc[~salary.State_Name.isin(['Puerto Rico'])].groupby(by='Region').Sample_Measurement.describe()